README.md
# domoticz-plugin-tinkerforge-lcd20x4-lite
[Domoticz](https://www.domoticz.com/) plugin to interact with the [Tinkerforge LCD 20x4 Bricklet](https://www.tinkerforge.com/en/doc/Hardware/Bricklets/LCD_20x4.html#lcd-20x4-bricklet).

# Objectives
* To write characters to the Tinkerforge LCD 20x4 Bricklet.
* To learn how to write Python plugin(s) for the Domoticz Home Automation system interacting with [Tinkerforge](http://www.tinkerforge.com) Building Blocks.

_Abbreviations_: TF=Tinkerforge, Bricklet=Tinkerforge LCD 20x4 Bricklet, GUI=Domoticz Web UI.

## Solution
A Domoticz Python plugin named **"Tinkerforge LCD 20x4 Bricklet - Lite"** enables to write characters to the LCD 20x4 display via a Domoticz text device (Type:General, SubType:Text).
At plugin start, an IP connection is established until either Domoticz stops or the plugin is disabled.
The text to be displayed on the LCD, is defined as a JSON formatted string (array with up-to 4 line items) hold by the Domoticz text device.
Custom characters are defined in an external file and displayed via Unicode. LCD table characters are also displayed via Unicode.
When the plugin starts, the backlight is switched on, the cursor is turned off and not blinking.
If the text of the Domoticz device is modified, the plugin connects via IP to Tinkerforge, writes the LCD line(s) as defined by the JSON string and disconnects.
Modifying the text of the Domoticz device can be done via scripts, API etc. See examples below.

**Note:**
This is the lite version of the plugin. The plugin full version enables to use the buttons and set the configuration for the backlight and cursor via switches.

### Domoticz Devices
* Type:General, SubType:Text, Name: JSON.

### Configuration
* Address: IP address of the host connected to. Default: 127.0.0.1 (for USB connection)
* Port: Port used by the host. Default: 4223
* UID: Unique identifier of the LCD 20x4 Bricklet. Obtain the UID via the Brick Viewer. Default: BHN

The bricklet is connected to a Tinkerforge Master Brick which is direct connected via USB with the Domoticz Home Automation system.
The Domoticz Home Automation system is running on a Raspberry Pi 3B+.

### Logic
The only function of the plugin is to set the text for each line to the display.
Any additional logic to be defined in Domoticz. Either by additional devices or scripts.
_Examples_ (Scripts)
* Clock - Display clock CHH:MM - C is the custom character (index1, \u0009) clock symbol (update every minute).
* Raspberry Pi Monitor - Display the CPU Temperature, RAM & Disc usage and the time (update every minute).
More see section "dzVents Lua Automation Script Examples".

## Hardware Parts
* Raspberry Pi 3B+ [(Info)](https://www.raspberrypi.org)
* Tinkerforge Master Brick 2.1 FW 2.4.10 [(Info)](https://www.tinkerforge.com/en/doc/Hardware/Bricks/Master_Brick.html#master-brick)
* Tinkerforge LCD 20x4 Bricklet 1.2 FW 2.0.6 [(Info)](https://www.tinkerforge.com/en/doc/Hardware/Bricklets/LCD_20x4.html#lcd-20x4-bricklet)

## Software
* Raspberry Pi Raspian Debian Linux Buster 4.19.93-v7+ #1290
* Domoticz Home Automation System V4.11717 (beta) 
* Tinkerforge Brick Daemon 2.4.1, Brick Viewer 2.4.11
* Tinkerforge Python API-Binding 2.1.24 [(Info)](https://www.tinkerforge.com/en/doc/Software/Bricklets/LCD20x4_Bricklet_Python.html)
* Python 3.7.3, GCC 8.2.0
* The versions for developing this plugin are subject to change.

## Quick Steps
For implementing the Plugin on the Domoticz Server running on the Raspberry Pi.
See also Python Plugin Code (well documented) **plugin.py**.

## Test Setup
For testing this plugin, the test setup has a Master Brick with LCD 20x4 Bricklet connected to port B.
On the Raspberry Pi, it is mandatory to install the Tinkerforge Brick Daemon and Brick Viewer following [these](https://www.tinkerforge.com/en/doc/Embedded/Raspberry_Pi.html) installation instructions (for Raspian armhf).
Start the Brick Viewer and action:
* Update the devices firmware
* Obtain the UID of the LCD 20x4 Bricklet as required by the plugin (i.e. BHN).

## Domoticz Web GUI
Open windows GUI Setup > Hardware, GUI Setup > Log, GUI Setup > Devices
This is required to add the new hardware with its device and monitor if the plugin code is running without errors.
It is imporant, that the GUI > Setup > Hardware accepts new devices!

## Create folder
The folder name is the same as the key property of the plugin (i.e. plugin key="TFLCD20X4").
```
cd /home/pi/domoticz/plugins/tflcd20x4lite
```

## Create the plugin
The Python plugin has a mandatory filename **plugin.py** located in the created plugin folder.
Domoticz Python Plugin Source Code: see file **plugin.py**.
In addition to the plugin file, the custom character file **customchar.json** is required.

### Plugin Pseudo Code
```
Define imports and amend path

Define constants for the devices (Units)

Define class BasePlugin:
	Init
		define debug flag
	onStart
		If first time, create the device Type General, SubType Text
		Configure the bricklet backlight on, cursor off, blink off = [function set_bricklet_configuration()]
	onDeviceModified(self, Unit)
		Handle text changes for the unit (the text device with JSON formatted string)
		UNITJSON = [function write_lines(unit)]
```

### Communication
Bricklet communication is on request = connect, execute function and disconnect from the Master Brick.
There is no ongoing connection between the plugin and Domoticz, as no callback used.
The plugin function onHeartbeat (every 10s) is not used.

### Write Lines
The text to be written to the display is defined in a JSON formatted string hold by the Domotiz Text device (default name Hardware - JSON, i.e. LCD20x4 - JSON).
The JSON string is an array with up-to 4 line items - at least 1 line must be defined.
```
[{"line":1,"position":n,"clear":n,"text":"Text"},{"line":2,"position":n,"clear":n,"text":"Text"},{"line":3,"position":n,"clear":n,"text":"Text"},{"line":4,"position":n,"clear":n,"text":"Text"}] 
```
Each line item has the **key:value pairs**:
"line":0-3 - Integer with line index 0 to 3.
"position":0-19 - Integer with position index 0 - 19.
"clear":1-2 - Integer to clear the line (1) or clear the display (2) prior writing the text to the line.
"Text":"Text string" - The text to be displayed.

#### Example Single Line
JSON string writing "Hello World" to line index 1 at position 9 and clearing the display prior writing.
```
[{"line":1,"position":9,"clear":2,"text":"Hello World"}] 
```

**Lua**
Define the JSON string and update the Domotiz device (General/Text,idx=84) which triggers writing to the LCD display.
```
IDXJSON = 84
local line = '[{"line":%d,"position":%d,"clear":2,"text":"%s"}]'
line = string.format(line, 1, 9, "Hello World" )
domoticz.devices(IDXJSON).updateText(line)
```

**API**
Send via HTTP API REST request, the JSON string for the device with idx=84.
This is a fast way to test the plugin.
In the URL, the special characters are escaped: "=%22, space=%20.
Do not embed the JSON array in "", i.e. "[{}]", but leave the "" out, i.e. [{}]
```
http://domoticz-ip-address:8080/json.htm?type=command&param=udevice&idx=84&nvalue=0&svalue=[{%22line%22:1,%22position%22:9,%22clear%22:2,%22text%22:%22Hello%20World!%22}]
```
with HTTP response:
```
{"status" : "OK","title" : "Update Device"}
```
and Domoticz Log:
```
2020-02-14 16:08:36.002 (LCD20x4) Pushing 'CPluginMessageBase' on to queue 
2020-02-14 16:08:36.015 (LCD20x4) Processing 'CPluginMessageBase' message 
2020-02-14 16:08:36.016 (LCD20x4) Calling message handler 'onDeviceModified'. 
2020-02-14 16:08:36.016 (LCD20x4) onDeviceModified called Unit:1 (LCD20x4 - JSON),nValue=0,sValue=[{"line":1,"position":9,"clear":2,"text":"Hello World!"}] 
2020-02-14 16:08:36.017 (LCD20x4) write_lines: Unit=1,ID=84,JSON=[{"line":1,"position":9,"clear":2,"text":"Hello World!"}] 
2020-02-14 16:08:36.029 (LCD20x4) write_lines: Lines: 1 
2020-02-14 16:08:36.030 (LCD20x4) write_lines: Line=1,Position=9,Text=Hello World! 
2020-02-14 16:08:36.134 (LCD20x4) write_lines: OK
```

#### Example All Lines
JSON string writing all 4 lines and clear the lines prior writing.
Basically some data to monitor the Raspberry Pi Domoticz System is displayed.
```
[
{"line":1,"position":0,"clear":1,"text":"CPU 43.5 \u00DFC"},
{"line":2,"position":0,"clear":1,"text":"RAM 29.42 %"},
{"line":3,"position":0,"clear":1,"text":"DISC 69.08 %"},
{"line":0,"position":14,"clear":1,"text":"\u000909:37"}
] 
```

### Custom Characters
Custom Characters are defined as an JSON array in an external file  located in the plugin folder.
The content is parsed and assigned to the Tinkerforge character set during plugin start.
Set Custom Characters with index 0-7 as max 8 custom characters can be defined.
JSON format example with some characters.
The id has a range 0-7, name is to know what character is defined, char is the 8 bit definition as decimals.
```
[
{"id":0,"name":"battery","char":"14,27,17,17,17,17,17,31"},
{"id":1,"name":"clock","char":"31,17,10,4,14,31,31,0"},
{"id":2,"name":"arrowdown","char":"0,4,4,4,31,14,4,0"},
{"id":3,"name":"arrowup","char":"0,4,14,31,4,4,4,0"},
{"id":4,"name":"nochange","char":"0,0,31,0,31,0,0,0"}
]
```

_Python_
Ensure to write in Unicode \u00NN and NOT \xNN.
Example using the device object:
lcdDev.write_line(0, 0, "Battery: " + "\u0008")

_Lua_
Ensure to escape the backslash \ character.
Example using string format to define a line with custom character clock, which has index 1(id:1) and unicode format \u0009:
local line0 = string.format('{"line":1,"position":0,"clear":1,"text":"Clock  %s  \\u0009"}', timenow )

### LCD Table Characters
To display a character from the LCD, use the same approach as for custom characters.
Example displaying the degree character, which is at position DF. Displayed is °C.
local line0 = string.format('{"line":1,"position":0,"clear":1,"text":"CPU  %s  \\u00DFC"}', cputemp )

### Hints
#### Clear Display
To clear the display, define a JSON string with empty text and clear = 2 at line = 0, position = 0.
```
[{"line":0,"position":0,"clear":2,"text":""}]
```

#### Bricklet Disconnected
If the LCD 20x4 bricklet is disconnected and connected again to the Domoticz system, the plugin must be updated (GUI > Setup > Hardware > select plugin).

## Install the Tinkerforge Python API
There are two options.

### 1) sudo pip3 install tinkerforge
Advantage: in case of binding updates, only a single folder must be updated.
Check if a subfolder tinkerforge is created in folder /usr/lib/python3/dist-packages.
**Note**
Check the version of "python3" in the folder path. This could also be python 3.7 or other = see below.

**If for some reason the bindings are not installed**
Unzip the Tinkerforge Python Binding into the folder /usr/lib/python3/dist-packages.
_Example_
Create subfolder Tinkerforge holding the Tinkerforge Python Library
```
cd /home/pi/tinkerforge
```
Unpack the latest python bindings into folder /home/pi/tinkerforge
Copy /home/pi/tinkerforge to the Python3 dist-packges
```
sudo cp -r /home/pi/tinkerforge /usr/lib/python3/dist-packages/
```

In the Python Plugin code amend the import path to enable using the Tinkerforge libraries
```
from os import path
import sys
sys.path
sys.path.append('/usr/local/lib/python3.7/dist-packages')
```

### 2) Install the Tinkerforge Python Bindings in a subfolder of the plugin and copy the binding content.
Disadvantage: Every Python plugin using the Tinkerforge bindings must have a subfolder tinkerforge.
In case of binding updates,each of the tinkerforge plugin folders must be updated.
/home/pi/domoticz/plugins/...

There is no need to amend the path as for option 1.

For either ways, the bindings are used like:
```
import tinkerforge
from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_lcd_20x4 import BrickletLCD20x4
```
**Note**
Add more bindings depending Tinkerforge bricks & bricklets used.

Ensure to update the files in case of newer Tinkerforge Python Bindings.

## Make plugin.py executable
```
cd /home/pi/domoticz/plugins/tflcd20x4lite
chmod +x plugin.py
```

## Restart Domoticz
Restart Domoticz to find the plugin:
```
sudo systemctl restart domoticz.service
```

**Note**
When making changes to the Python plugin code, ensure to restart Domoticz and refresh any of the Domoticz Web GUI's.

## Domoticz Add Hardware Tinkerforge LCD 20x4 Bricklet
**IMPORTANT**
Prior adding, set GUI Stup > Settings > Hardware the option to allow new hardware.
If this option is not enabled, no new devices are created assigned to this hardware.
Check in the Domoticz log as error message Python script at the line where the new device is used
(i.e. Domoticz.Debug("Device created: "+Devices[1].Name))

In the GUI Setup > Hardware add the new hardware "Tinkerforge LCD 20x4 Bricklet - Lite".

## Add Hardware - Check the Domoticz Log
After adding,ensure to check the Domoticz Log (GUI Setup > Log)

## Example Domoticz Log Entry Adding Hardware with Debug=True
```
2020-02-14 10:23:49.483 (LCD20x4) Debug logging mask set to: PYTHON PLUGIN QUEUE IMAGE DEVICE CONNECTION MESSAGE ALL 
2020-02-14 10:23:49.483 (LCD20x4) 'HardwareID':'10' 
2020-02-14 10:23:49.483 (LCD20x4) 'HomeFolder':'/home/pi/domoticz/plugins/tflcd20x4lite/' 
2020-02-14 10:23:49.483 (LCD20x4) 'StartupFolder':'/home/pi/domoticz/' 
2020-02-14 10:23:49.483 (LCD20x4) 'UserDataFolder':'/home/pi/domoticz/' 
2020-02-14 10:23:49.483 (LCD20x4) 'Database':'/home/pi/domoticz/domoticz.db' 
2020-02-14 10:23:49.483 (LCD20x4) 'Language':'en' 
2020-02-14 10:23:49.483 (LCD20x4) 'Version':'1.0.0' 
2020-02-14 10:23:49.483 (LCD20x4) 'Author':'rwbL' 
2020-02-14 10:23:49.483 (LCD20x4) 'Name':'LCD20x4' 
2020-02-14 10:23:49.483 (LCD20x4) 'Address':'127.0.0.1' 
2020-02-14 10:23:49.484 (LCD20x4) 'Port':'4223' 
2020-02-14 10:23:49.484 (LCD20x4) 'Key':'tflcd20x4lite' 
2020-02-14 10:23:49.484 (LCD20x4) 'Mode1':'BHN' 
2020-02-14 10:23:49.484 (LCD20x4) 'Mode6':'Debug' 
2020-02-14 10:23:49.484 (LCD20x4) 'DomoticzVersion':'4.11674' 
2020-02-14 10:23:49.484 (LCD20x4) 'DomoticzHash':'91c3475ac' 
2020-02-14 10:23:49.484 (LCD20x4) 'DomoticzBuildTime':'2020-02-09 10:00:09' 
2020-02-14 10:23:49.484 (LCD20x4) Device count: 0 
2020-02-14 10:23:49.484 (LCD20x4) Creating new Device(s) 
2020-02-14 10:23:49.484 (LCD20x4) Creating device 'JSON'. 
2020-02-14 10:23:49.485 (LCD20x4) Device created: LCD20x4 - JSON 
2020-02-14 10:23:49.486 (LCD20x4) set_bricklet_configuration 
2020-02-14 10:23:49.490 (LCD20x4) Customchar: #characters defined: 5 
2020-02-14 10:23:49.491 (LCD20x4) Customchar: Index=0,Name=battery,Char=14,27,17,17,17,17,17,31 
2020-02-14 10:23:49.491 (LCD20x4) Customchar: Index=1,Name=clock,Char=31,17,10,4,14,31,31,0 
2020-02-14 10:23:49.491 (LCD20x4) Customchar: Index=2,Name=arrowdown,Char=0,4,4,4,31,14,4,0 
2020-02-14 10:23:49.491 (LCD20x4) Customchar: Index=3,Name=arrowup,Char=0,4,14,31,4,4,4,0 
2020-02-14 10:23:49.492 (LCD20x4) Customchar: Index=4,Name=nochange,Char=0,0,31,0,31,0,0,0 
2020-02-14 10:23:49.593 (LCD20x4) set_bricklet_configuration: OK 
2020-02-14 10:23:49.020 Status: (LCD20x4) Started. 
2020-02-14 10:23:49.480 Status: (LCD20x4) Initialized version 1.0.0, author 'rwbL' 
2020-02-14 10:23:49.480 Status: (LCD20x4) Entering work loop.
```

## dzVents Lua Automation Script Examples

### Hello World
```
-- Tinkerforge LCD20x4  Bricklet Plugin - Test Script
-- dzVents Automation Script: tflcd20x4_hello_world
-- Display the text "Hello World".
-- A JSON array with a single entry is used to set the text.
-- To avoid spoiling the log of the text device, the log is cleared at every time update - using function openURL
-- 20200114 by rwbL

-- Idx of the text device for JSON text.
IDXJSON = 84    -- LCD line&position set by JSON string array

return {
	on = {
		timer = {
			'every minute',
	   },
    },
	execute = function(domoticz, timer)
        local line = '[' ..
                        '{"line":%d,"position":%d,"clear":2,"text":"%s"}' ..
                     ']'
        line = string.format(line, 1, 9, "Hello World" )
        domoticz.devices(IDXJSON).updateText(line)
        -- Clear the device log
        domoticz.openURL('http://localhost:8080/json.htm?type=command&param=clearlightlog&idx=' .. IDXJSON)
    end
}
```

### Clock
```
-- Tinkerforge LCD20x4  Bricklet Plugin - Test Script Clock
-- dzVents Automation Script: tflcd20x4_clock
-- Display clock CHH:MM - C is the custom character (index1, \u0009) clock symbol.
-- The time, updated every minute, position is randomly on the LCD display at one of the 4 lines 0-3 and at positions 0-13 (=19 - length time characters 6).
-- A JSON array with a single entry is used to set the time.
-- To avoid spoiling the log of the text device, the log is cleared at every time update - using function openURL
-- 20200109 by rwbL


CHANGELOG.md
# Changelog domoticz-plugin-tinkerforge-lcd20x4-lite

### v1.0.0 (Build 20200214)
* NEW: First version published.


plugin.py
# Domoticz Home Automation - Plugin Tinkerforge LCD 20x4 Bricklet - Lite
# @author Robert W.B. Linn
# @version 1.0.0 (Build 20200214)
#
# NOTE: after every change restart Domoticz & check the Domoticz Log
# sudo systemctl restart domoticz.service OR sudo service domoticz.sh restart
# REFERENCES:
# Domoticz Python Plugin Development:
# https://www.domoticz.com/wiki/Developing_a_Python_plugin
# Tinkerforge LCD 20x4 Bricklet:
# Hardware:
# https://www.tinkerforge.com/en/doc/Hardware/Bricklets/LCD_20x4.html#lcd-20x4-bricklet
# API Python Documentation:
# https://www.tinkerforge.com/en/doc/Software/Bricklets/LCD20x4_Bricklet_Python.html

"""
<plugin key="tflcd20x4lite" name="Tinkerforge LCD 20x4 Bricklet - Lite" author="rwbL" version="1.0.0">
    <description>
        <h2>Tinkerforge LCD 20x4 Bricklet - Lite</h2><br/>
        The plugin enables to write characters to the LCD 20x4 display via a Domoticz text device (Type:General, SubType:Text).<br/>
        The text to be displayed is defined as a JSON formatted string (array with up-to 4 line items).<br/>
        [{"line":1,"position":n,"clear":n,"text":"Text"},{"line":2,"position":n,"clear":n,"text":"Text"},{"line":3,"position":n,"clear":n,"text":"Text"},{"line":4,"position":n,"clear":n,"text":"Text"}] <br/>
        Each line item has the key:value pairs:
        <ul style="list-style-type:square">
            <li>"line":0-3 - Integer with line index 0 to 3.</li>
            <li>"position":0-19 - Integer with position index 0 - 19.</li>
            <li>"clear":1-2 - Integer to clear the line (1) or clear the display (2) prior writing the text to the line.</li>
            <li>"Text":"Text string" - The text to be displayed.</li>
        </ul>
        Custom characters are defined in an external file and displayed via Unicode.<br/>
        When the plugin starts, the backlight is switched on, the cursor is turned off and not blinking.<br/>
        If the text of the Domoticz device is modified, the plugin connects via IP to Tinkerforge, writes the LCD line(s) as defined by the JSON string and disconnects.<br/>
        <br/>
        Note: This is the lite version of the plugin. The full version enables to use the buttons and set the configuration for the backlight and cursor.<br/>
        <br/>
        <h3>Domoticz Devices</h3>
        <ul style="list-style-type:square">
            <li>Type:General, SubType:Text, Name: JSON.</li>
        </ul>
        <br/>
        <h3>Configuration</h3>
        <ul style="list-style-type:square">
            <li>Address: IP address of the host connected to. Default: 127.0.0.1 (for USB connection)</li>
            <li>Port: Port used by the host. Default: 4223</li>
            <li>UID: Unique identifier of the LCD 20x4 Bricklet. Obtain the UID via the Brick Viewer. Default: BHN</li>
        </ul>
    </description>
    <params>
        <param field="Address" label="Host" width="200px" required="true" default="127.0.0.1"/>
        <param field="Port" label="Port" width="75px" required="true" default="4223"/>
        <param field="Mode1" label="UID" width="200px" required="true" default="BHN"/>
        <param field="Mode6" label="Debug" width="75px">
            <options>
                <option label="True" value="Debug" default="true"/>
                <option label="False" value="Normal"/>
            </options>
        </param>
    </params>
</plugin>
""" 

## Imports
import Domoticz
import urllib
import urllib.request
import json 

# Amend the import path to enable using the Tinkerforge libraries
# Alternate (ensure to update in case newer Python API bindings):
# create folder tinkerforge and copy the binding content, i.e.
# /home/pi/domoticz/plugins/tflcd20x4lite
from os import path
import sys
sys.path
sys.path.append('/usr/local/lib/python3.7/dist-packages')

import tinkerforge
from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_lcd_20x4 import BrickletLCD20x4

# Units
## Text lines set using JSON format - UNIT MUST start with 1
UNITJSON = 1
# Status Level & Text
STATUSLEVELOK = 1
STATUSLEVELERROR = 5
STATUSTEXTOK = "OK"
STATUSTEXTERROR = "ERROR"

# Custom Characters file JSON array format - located in the same folder as plugin.py. older name is plugin key.
CUSTOMCHARFILE = "/home/pi/domoticz/plugins/tflcd20x4lite/customchar.json"

# Messages (not all)
MSGERRNOUID = "[ERROR] Bricklet UID not set. Get the UID using the Brick Viewer."
MSGERRSETCONFIG = "[ERROR] Set bricklet configuration failed. Check bricklet and settings."

class BasePlugin:

    def __init__(self):
        self.Debug = False
        
        # NOT USED=Placeholder
        # The Domoticz heartbeat is set to every 10 seconds. Do not use a higher value than 30 as Domoticz message "Error: hardware (N) thread seems to have ended unexpectedly"
        # The plugin heartbeat is set in Parameter.Mode5 (seconds). This is determined by using a hearbeatcounter which is triggered by:
        # (self.HeartbeatCounter * self.HeartbeatInterval) % int(Parameter.Mode5) = 0
        # Example: trigger action every 60s [every 6 heartbeats] = (6 * 10) mod 60 = 0 or every 5 minutes (300s) [every 30 heartbeats] = (30 * 10) mod 300 = 0
        # self.HeartbeatInterval = 10
        # self.HeartbeatCounter = 0

    def onStart(self):
        Domoticz.Debug("onStart called")
        Domoticz.Debug("Debug Mode:" + Parameters["Mode6"])
        if Parameters["Mode6"] == "Debug":
            self.debug = True
            Domoticz.Debugging(1)
            dump_config_to_log()

        if (len(Devices) == 0):
            # Create new devices for the Hardware
            Domoticz.Debug("Creating new Device(s)")
            Domoticz.Device(Name="JSON", Unit=UNITJSON, TypeName="Text", Used=1).Create()
            Domoticz.Debug("Device created: "+Devices[UNITJSON].Name)

        # Get the UID of the bricklet
        if len(Parameters["Mode1"]) == 0:
            status_to_log(STATUSLEVELERROR, MSGERRNOUID)
            return

        # Set the bricklet configuration: backlight on,cursor on, blink on
        set_configuration(True,False,False)

    def onStop(self):
        Domoticz.Debug("Plugin is stopping.")

    def onConnect(self, Connection, Status, Description):
        Domoticz.Debug("onConnect called")

    def onMessage(self, Connection, Data):
        Domoticz.Debug("onMessage called")

    def onCommand(self, Unit, Command, Level, Hue):
        Domoticz.Debug("onCommand called Unit " + str(Unit) + ": Parameter '" + str(Command) + "', Level: " + str(Level) + "', Hue: " + str(Hue))

    # Handle new text from the text units (devices). The text is in property Devices[Unit].sValue
    def onDeviceModified(self, Unit):
        Domoticz.Debug("onDeviceModified called Unit:" + str(Unit) + " (" + Devices[Unit].Name + "),nValue="+str(Devices[Unit].nValue) + ",sValue="+Devices[Unit].sValue)
        write_lines(Unit)
        
    def onNotification(self, Name, Subject, Text, Status, Priority, Sound, ImageFile):
        Domoticz.Debug("Notification: " + Name + "," + Subject + "," + Text + "," + Status + "," + str(Priority) + "," + Sound + "," + ImageFile)

    def onDisconnect(self, Connection):
        Domoticz.Debug("onDisconnect called")

    def onHeartbeat(self):
        Domoticz.Debug("onHeartbeat called")
        # NOT USED = PLACEHOLDER
        """
        self.HeartbeatCounter = self.HeartbeatCounter + 1
        Domoticz.Debug("onHeartbeat called. Counter=" + str(self.HeartbeatCounter * self.HeartbeatInterval) + " (Heartbeat=" + Parameters["Mode5"] + ")")
        # check the heartbeatcounter against the heartbeatinterval
        if (self.HeartbeatCounter * self.HeartbeatInterval) % int(Parameters["Mode5"]) == 0:
            try:
                SetBrickletIlluminance(UNITILLUMINATION)
                return
            except:
                #Domoticz.Error("[ERROR] ...")
                return
        """

global _plugin
_plugin = BasePlugin()

def onStart():
    global _plugin
    _plugin.onStart()

def onStop():
    global _plugin
    _plugin.onStop()

def onConnect(Connection, Status, Description):
    global _plugin
    _plugin.onConnect(Connection, Status, Description)

def onMessage(Connection, Data):
    global _plugin
    _plugin.onMessage(Connection, Data)

def onCommand(Unit, Command, Level, Hue):
    global _plugin
    _plugin.onCommand(Unit, Command, Level, Hue)

def onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile):
    global _plugin
    _plugin.onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile)

def onDisconnect(Connection):
    global _plugin
    _plugin.onDisconnect(Connection)

def onDeviceModified(Unit):
    global _plugin
    _plugin.onDeviceModified(Unit)

def onHeartbeat():
    global _plugin
    _plugin.onHeartbeat()

# Tinkerforge Bricklet

"""
Set the bricklet configuration
Parameter:
backlight - True|False = Set backlight on
cursor - True|false = set cursor on|off
blick - True|False = set blink cursor on|off
"""
def set_configuration(backlight,cursor,blinking):
    Domoticz.Debug("set_configuration")
    try:
        # Create IP connection
        ipConn = IPConnection()
        # Create device object
        lcdDev = BrickletLCD20x4(Parameters["Mode1"], ipConn)
        # Connect to brickd using Host and Port
        ipConn.connect(Parameters["Address"], int(Parameters["Port"]))
        # Update the configuration
        if backlight == True:
            lcdDev.backlight_on()
        else:
            lcdDev.backlight_off()
        lcdDev.set_config(cursor, blinking)
        
        # Set Custom Characters with index 0-7 as max 8 custom characters can be defined.
        # dzVents Lua scripts:
        # The \ character needs to be escaped, i.e. line = string.format(line, l, p, "\\u0008", domoticz.helpers.isnowhhmm(domoticz) )
        # Ensure to write in unicode \u00NN and NOT \xNN - Example: lcdDev.write_line(0, 0, "Battery: " + "\u0008")
        # JSON has no hex escape \xNN but supports unicode escape \uNNNN
        # 
        # Example custom character definition and write to the lcd:
        #     battery = [14,27,17,17,17,17,17,31] 
        #     clock = [31,17,10,4,14,31,31,0]
        #     lcdDev.set_custom_character(0, clock)
        # 
        # JSON File
        # Read the custom characters from JSON array file as defined by constant CUSTOMCHARFILE
        # JSON format examplewith 2 characters: 
        # [
        #     {"id":0,"name":"battery","char":"14,27,17,17,17,17,17,31"},
        #     {"id":1,"name":"clock","char":"31,17,10,4,14,31,31,0"}
        # ]
        # Use exception handling in case file not found
        # The id must be in range 0-7 as max 8 custom characters can be defined
        try:
            with open(CUSTOMCHARFILE) as f:
                json_char_array = json.load(f)
                Domoticz.Debug("Customchar: #characters defined: %d" % (len(json_char_array)) )
                if len(json_char_array) > 0:
                    for item in json_char_array:
                        id = int(item["id"])
                        name = item["name"]
                        char = item["char"].strip().split(",")
                        # Check if the character id is in range 0-7
                        if id >= 0 and id <= 7:
                            lcdDev.set_custom_character(id, char)
                            Domoticz.Debug("Customchar: Index=%d,Name=%s,Char=%s" % (id,name,item["char"]) )
                        else:
                            status_to_log(STATUSLEVELERROR,"Customchar: Index=%d not in range 0-7." % (id) )
                else:
                    status_to_log(STATUSLEVELERROR,"Customchar: No or wrong characters defined.")
        except:
            status_to_log(STATUSLEVELERROR,"Customchar: Can not open file=%s." % (CUSTOMCHARFILE) )
        # Disconnect
        ipConn.disconnect()
        Domoticz.Debug("set_configuration: OK")
    except:
        status_to_log(STATUSLEVELERROR, MSGERRSETCONFIG)
    return

"""
Set LCD text for the 4 lines and 20 columns using json.
Parameter:
unit - unit number as defined in the constants (see top) for the text device holding the JSON string.
The text as JSON array string [{},{}] is provided by a device unit using the property device.sValue
The JSON array has 1 to N entries with line properties.
It is possible to define an entry for a line or multiple entries for a line for different positions.
JSON keys with value explain
"line" - 0-3
"position" - 0-19
"clear": clear the display: 0=no clear,1=clear line,2=clear display
"text": text to display at the line

Example JSON string = JSON array for the 4 lines
jsonarraystring = '
[
    {"line":0,"position":0,"clear":1,"text":"TEXT 1"},
    {"line":1,"position":0,"clear":1,"text":"TEXT 2"},
    {"line":2,"position":0,"clear":1,"text":"TEXT 3"},
    {"line":3,"position":0,"clear":1,"text":"TEXT 4"}
]'
"""
def write_lines(unit):
    Domoticz.Debug("write_lines: Unit=%d,ID=%d,JSON=%s" % (unit, Devices[unit].ID, Devices[unit].sValue) )
    EMPTYLINE = "                    "
    jsonstring = Devices[unit].sValue
    if len(jsonstring) == 0:
        status_to_log(STATUSLEVELERROR,"No JSON string defined for the text device (Unit sValue empty).")
        return
    try:
        # Create IP connection
        ipConn = IPConnection()
        # Create device object
        lcdDev = BrickletLCD20x4(Parameters["Mode1"], ipConn)
        # Connect to brickd using Host and Port
        ipConn.connect(Parameters["Address"], int(Parameters["Port"]))
        # Define the lcd function write_line parameter
        line = 0
        position = 0
        text = ""
        clear = 0
        # parse the json string
        json_array = json.loads(jsonstring, encoding=None)
        Domoticz.Debug("write_lines: Lines: %d" % (len(json_array)) )
        for item in json_array:
            line = int(item["line"])
            position = int(item["position"])
            text = item["text"]
            clear = int(item["clear"])
            # Domoticz.Debug("write_lines: Items (l,p,t,c): %d,%d,%s,%d" % (line,position,text,clear) )
            # Checks
            if (line < 0 or line > 3):
                status_to_log(STATUSLEVELERROR, "write_lines: Wrong line number: %d. Ensure 0-3." % (line) )
                return
            if (position < 0 or position > 19):
                status_to_log(STATUSLEVELERROR, "write_lines: Wrong position number: %d. Ensure 0-19." % (position) )
                return
            # Clear action: 1=clear line;2=clear display
            if clear == 1:
                lcdDev.write_line(line, 0, EMPTYLINE)
            if clear == 2:
                lcdDev.clear_display()
            # Write text 
            lcdDev.write_line(line, position, text)
            status_to_log(STATUSLEVELOK, "write_lines: Line=%d,Position=%d,Text=%s" % (line,position,text) )
        # Disconnect
        ipConn.disconnect()
        Domoticz.Debug("write_lines: OK")
    except:
        status_to_log(STATUSLEVELERROR, "write_lines: Failed writing text (Unit=%d,ID=%d,JSON=%s). Check JSON definition." % (unit, Devices[unit].ID, Devices[unit].sValue) )
    return

##
# Generic helper functions
##

# Dump the plugin parameter & device information to the domoticz debug log
def dump_config_to_log():
    for x in Parameters:
        if Parameters[x] != "":
            Domoticz.Debug( "'" + x + "':'" + str(Parameters[x]) + "'")
    Domoticz.Debug("Device count: " + str(len(Devices)))
    for x in Devices:
        Domoticz.Debug("Device:           " + str(x) + " - " + str(Devices[x]))
        Domoticz.Debug("Device ID:       '" + str(Devices[x].ID) + "'")
        Domoticz.Debug("Device Name:     '" + Devices[x].Name + "'")
        Domoticz.Debug("Device nValue:    " + str(Devices[x].nValue))
        Domoticz.Debug("Device sValue:   '" + Devices[x].sValue + "'")
        Domoticz.Debug("Device LastLevel: " + str(Devices[x].LastLevel))
    return

# Log a status message, either to the domoticz log or error
# Status level: 1=OK, 4=ERROR
def status_to_log(level,text):
    if level == STATUSLEVELOK:
        Domoticz.Log(text)
    if level == STATUSLEVELERROR:
        Domoticz.Error(text)
    return
  
# Check if a string is JSON format  
def string_is_json(myjson):
    try:
        json_object = json.loads(myjson)
    except ValueError as e:
        return False
    return True
